io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -98,6 +124,40 @@ selectTraceEvents' fn =
               []
     . traceSelectTraceEvents fn
 
+selectTraceRaces :: SimTrace a -> [ScheduleControl]
+selectTraceRaces = go
+  where
+    go (SimTrace _ _ _ _ trace)      = go trace
+    go (TraceRacesFound races trace) =
+      races ++ go trace
+    go _                             = []
+
+-- Extracting races from a trace.  There is a subtlety in doing so: we
+-- must return a defined list of races even in the case where the
+-- trace is infinite, and there are no races occurring in it! For
+-- example, if the system falls into a deterministic infinite loop,
+-- then there will be no races to find.
+
+-- In reality we only want to extract races from *the part of the
+-- trace used in a test*. We can only observe that by tracking lazy
+-- evaluation: only races that were found in the evaluated prefix of
+-- an infinite trace should contribute to the "races found". Hence we
+-- return a function that returns the races found "so far". This is
+-- unsafe, of course, since that function may return different results
+-- at different times.
+
+detachTraceRaces :: SimTrace a -> (() -> [ScheduleControl], SimTrace a)
+detachTraceRaces trace = unsafePerformIO $ do
+  races <- newIORef []
+  let readRaces ()  = concat . reverse . unsafePerformIO $ readIORef races
+      saveRaces r t = unsafePerformIO $ do
+                        modifyIORef races (r:)
+                        return t
+  let go (SimTrace a b c d trace)  = SimTrace a b c d $ go trace
+      go (TraceRacesFound r trace) = saveRaces r   $ go trace
+      go t                         = t
+  return (readRaces,go trace)
```

Comment:
I don't think this can work. I'm pretty sure programs using ST are deterministic, and the results can't depend on evaluation order--which is exactly what we need here. I wrote a little test program
[Test.hs.txt](https://github.com/input-output-hk/ouroboros-network/files/7959748/Test.hs.txt)
When I run the test, I get
*Main> test
([1,4,9,16,25,36,49,64,81,100],
and then ghci hangs--it's in a loop running go to 'completion', trying to do all the possible updates to ns. Even though we're using lazy ST, when we evaluate the result of readIORef ns, we force all the computation up to that point. detachTraceRaces works in exactly the same way.

Reply:
What about the following modification, which is

* using `unsafeInterleaveST` to read the `STRef`
* using a bang pattern to make sure that the `modifySTRef` is executed (that's the tricky part which I missed at first).

```hs
{-# LANGUAGE BangPatterns #-}
module Test where

import Data.STRef.Lazy

import Control.Monad.ST.Lazy
import Control.Monad.ST.Lazy.Unsafe

run :: ST s ([Integer],[Integer])
run = do
  ns <- newSTRef []
  n <- unsafeInterleaveST $ readSTRef ns
  xs <- go ns 1
  return (xs, n)

  where go ns i = do
          -- must be strict, otherwise the `readSTRef` will not see the alterations
          !_ <- modifySTRef ns (i:)
          ((i*i):) <$> go ns (i+1)

test = let (xs, n) = runST run in (take 10 xs, n)
```

The output in `ghci` is
```
([1,4,9,16,25,36,49,64,81,100],[10,9,8,7,6,5,4,3,2,1])
```

Reply:
I see, yes, that looks much better.

I'm still nervous about the change though. Getting the correct behaviour relies on n being evaluated at exactly the right time--or, in the real code, readRaces. I know that GHC can and does move the evaluation of variables earlier, if strictness analysis says the variable will eventually be evaluated. That's why I made readRaces into a function of ()... because there is no advantage in moving a function call earlier, so GHC is much less likely to do that. I don't think you can turn n into a function of () in the code above, because unsafeInterleaveST still returns an ST. In IOSim, readRaces is bound outside a .&&|, and used in the second argument, and must not be called until evaluation of the first argument is complete. Probably, even if it were not a function, the strictness analyser would not move the evaluation of readRaces earlier, and the code would still work. Probably. Will that be true of future versions of GHC? I honestly don't know.

So my fear is that this change would make the code more brittle, more likely to break on future versions of GHC. It is unsafe code, and fundamentally has to be unsafe, but that does mean we have to be very careful that compiler optimisations don't break it. Is the gain worth the risk?

Reply:
One can always force `n` to be lazy with
```hs
data Lazy a = Lazy {-# NOUNPACK #-} a
```
it would be highly dubious from ghc to remove `Lazy` altogether.

Even better option, is to use [lazy](https://hackage.haskell.org/package/base-4.16.0.0/docs/GHC-Exts.html#v:lazy) which restrains the strictness analysis.  If `lazy a` would become strict, also `par` would be broken (at least as currently implemented), which is unlikely to happen.